distributed/utils_comm.py

```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
                         address,
                         who=who,
                         serializers=serializers,
-                        max_connections=False,
                     ),
                     operation="get_data_from_worker",
                 ),
                 name=f"get-data-from-{address}",
             )
             for address, keys in d.items()
         }
-        response: dict[str, object] = {}
-        for worker, c in coroutines.items():
+        for address, task in tasks.items():
             try:
-                r = await c
-            except OSError:
-                missing_workers.add(worker)
-            except ValueError as e:
-                logger.info(
-                    "Got an unexpected error while collecting from workers: %s", e
+                r = await task
+            except (OSError, asyncio.CancelledError, asyncio.TimeoutError):
+                # Note: CancelledError and asyncio.TimeoutError are rare conditions
+                # that can be raised by the network stack.
+                # See https://github.com/dask/distributed/issues/8006
+                missing_workers.add(address)
```

Comment:
Catching this `CancelledError` is actually slightly worse than https://github.com/dask/distributed/pull/8013
In the other PR we were talking about a well isolated task and ensured that this task could still terminate safely when it caught a CancelledError. In this case, `gather_data_from_workers` is embedded in another coroutine and a CancelledError would actually not cause this stack to unravel but would instead cause a chain of events to handle `missing_workers`.